codeowners: enrich CODEOWNERS with team info and a debugging tool

otan opened the pull request:
Expand the CODEOWNERS functionality by adding a TEAMS.yaml file which
contains useful metadata about each given team.

The changes are as follows:
* Add a `team` package, which infers team information from a TEAMS.yaml
  file in root. This TEAMS.yaml file is enriched with a few more teams.
* Add a `codeowners` package which can interpret CODEOWNERS files and
  match a filepath to a set of teams which own a given file.
* Add a `whoownsit` tool which given any number of files or directories
  attempts to decipher the owner from CODEOWNERS.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/52053)
<!-- Reviewable:end -->

tbg commented on pkg/internal/codeowners/codeowners.go:
```diff
@@ -0,0 +1,143 @@
+
+func init() {
+	DefaultCodeOwnersLocation = filepath.Join(
+		os.Getenv("GOPATH"),
```
GOPATH isn't necessarily set, need to fall back to `os.ExpandEnv("$HOME/go")`

tbg commented on pkg/internal/codeowners/codeowners.go:
```diff
@@ -0,0 +1,143 @@
+	"github.com/zabawaba99/go-gitignore"
+)
+
+// DefaultCodeOwnersLocation is the default location for the TEAMS.yaml file.https://github.com/mattn/go-zglob
```
pasta

tbg reviewed (commented):
This looks good to me, as long as everyone accepts that the structure of CODOWNERs is not up to the individual teams any more (i.e. no more bulk-prs).

When there are multiple owners for a given file, are there precedence rules?

Are we expecting to want to lint against unowned files (guess yes!) and if so, how long is a typical CODEOWNERs invocation - can we add it as an advisory pre-push hook? (Not the kind that fails the push, but print some yellow lines). BTW not asking you to do any of this! We need to figure out if this is the right approach first.

otan committed:
codeowners: enrich CODEOWNERS with team info and a debugging tool

Expand the CODEOWNERS functionality by adding a TEAMS.yaml file which
contains useful metadata about each given team.

The changes are as follows:
* Add a `team` package, which infers team information from a TEAMS.yaml
  file in root. This TEAMS.yaml file is enriched with a few more teams.
* Add a `codeowners` package which can interpret CODEOWNERS files and
  match a filepath to a set of teams which own a given file.
* Add a `whoownsit` tool which given any number of files or directories
  attempts to decipher the owner from CODEOWNERS.

Release note: None

otan commented:
ok, i've reverted CODEOWNERS to dupe relevant PR owners for now. I'll make a change to remove that after some circulation.

otan commented:
bors r=tbg
